### PR TITLE
Gossip Fraud Proofs to P2P network after generation when needed

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
## Overview

Gossip generated Fraud Proofs to P2P network

Closes: #552 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

**Review by tzdybal — CHANGES_REQUESTED** [review]
Now I noticed, that the entire gossiping flow is invalid.
`newFraudProofValidator` should be kind of similar to `newTxValidator` - we should verify the fraud proof, and only if it was a valid fraud proof, we should propagate it (return true in validation function returned from `newFraudProofValidator`).

Right now, we always return `true`, so all fraud proofs are propagated in gossip network, and we also re-publish them in `fraudProofPublishLoop`. 

I think we don't need this loop at all - the only proof we need to "publish" is the one we generated, and currently our policy is to stop after generating.

File: state/executor.go [hunk 1]
```diff
@@ -47,6 +49,7 @@ func NewBlockExecutor(proposerAddress []byte, namespaceID [8]byte, chainID strin
 		fraudProofsEnabled: fraudProofsEnabled,
 		eventBus:           eventBus,
 		logger:             logger,
+		FraudProofOutCh:    make(chan *abci.FraudProof, 100),
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply state/executor.go:52]: Or, we even don't need this at all :thinking:

**Comment by Manav-Aggarwal** [comment]
Follow up issue: #693

**Review by nashqueue — APPROVED** [review]

File: block/manager.go [hunk 2]
```diff
@@ -244,24 +249,33 @@ func (m *Manager) SyncLoop(ctx context.Context) {
 			m.retrieveCond.Signal()
 
 			err := m.trySyncNextBlock(ctx, daHeight)
+			if err != nil && err.Error() == fmt.Errorf("failed to ApplyBlock: %w", state.ErrFraudProofGenerated).Error() {
```
Comment (tzdybal) [inline block/manager.go:252]: Create `ErrApplyBlock` with `Unwrap` method.

File: node/full.go [hunk 3]
```diff
@@ -217,6 +222,27 @@ func (n *FullNode) headerPublishLoop(ctx context.Context) {
...
 }
 
+func (n *FullNode) fraudProofPublishLoop(ctx context.Context) {
+	for {
+		select {
+		case fraudProof := <-n.blockManager.GetFraudProofOutChan():
+			n.Logger.Info("generated fraud proof: ", fraudProof.String())
+			fraudProofBytes, err := fraudProof.Marshal()
+			if err != nil {
+				n.Logger.Error("failed to serialize fraud proof", "error", err)
```
Comment (tzdybal) [inline node/full.go:232]: I know this shouldn't happen, but in case of an error, we shouldn't gossip.

**Review by tzdybal — CHANGES_REQUESTED** [review]
I left few comments. Refactoring error type can be done as followup issue.

File: block/manager.go (hunk 2 above)
Comment (tzdybal) [reply block/manager.go:252]: Follow up issue: https://github.com/rollkit/rollkit/issues/707

**Review by nashqueue — DISMISSED** [review]
LGTM, except that we should not gossip if we get an error from the comment above.

File: node/full.go (hunk 3 above)
Comment (Manav-Aggarwal) [reply node/full.go:232]: fixed

**Review by tzdybal — APPROVED** [review]
LGTM

**Review by nashqueue — APPROVED** [review]
LGTM

**Pull request merged**